File: extensions/vscode/src/test/telemetry.test.ts
import test from "node:test";
import assert from "node:assert/strict";
import {
  isTelemetryEnabled,
  buildMetricsEvent,
  compareVersions,
  evaluateVersionGate,
} from "../telemetry";

const FORBIDDEN_KEYS = [
  "raw_output", "raw_log", "prompt", "ai_reply", "command",
  "cwd", "path", "file_name", "repo_name", "username", "email",
  "api_key", "token", "full_session_id", "full_host_instance_hash",
  "full_workspace_hash",
  // Multi-channel attribution ids are LOCAL-only and must never be uploaded.
  "channel_id", "run_id", "turn_id", "event_id",
  "full_channel_id", "full_run_id", "full_turn_id",
];

test("VS Code global telemetry OFF disables sending (absolute)", () => {
  // Even if every other signal says "on", vscode-level off wins.
  assert.equal(
    isTelemetryEnabled({ vscodeTelemetryEnabled: false, xitSetting: "on", envOverride: "on" }),
    false
  );
});

test("default is enabled when vscode telemetry is on", () => {
  assert.equal(isTelemetryEnabled({ vscodeTelemetryEnabled: true, xitSetting: "default" }), true);
});

test("xit.telemetry=off disables sending", () => {
  assert.equal(isTelemetryEnabled({ vscodeTelemetryEnabled: true, xitSetting: "off" }), false);
});

test("XIT_TELEMETRY=off env overrides xit.telemetry=on", () => {
  assert.equal(
    isTelemetryEnabled({ vscodeTelemetryEnabled: true, xitSetting: "on", envOverride: "off" }),
    false
  );
});

test("XIT_TELEMETRY=on enables when vscode telemetry on", () => {
  assert.equal(
    isTelemetryEnabled({ vscodeTelemetryEnabled: true, xitSetting: "default", envOverride: "on" }),
    true
  );
});

test("built event contains no forbidden/sensitive fields", () => {
  const ev = buildMetricsEvent({
    installId: "anon123",
    vscodeExtensionVersion: "0.0.35",
    adapter: "claude",
    surface: "bridge",
    inputBytes: 100000,
    summaryBytes: 1000,
    savedBytes: 99000,
    runCount: 2,
    status: "success",
  });
  const keys = Object.keys(ev);
  for (const k of FORBIDDEN_KEYS) {
    assert.equal(keys.includes(k), false, `forbidden key present: ${k}`);
  }
});

test("saved-token estimate uses saved_bytes/4", () => {
  const ev = buildMetricsEvent({
    installId: "x",
    vscodeExtensionVersion: "0.0.35",
    savedBytes: 122222,
    inputBytes: 130000,
  });
  assert.equal(ev.estimated_saved_tokens, 30555);
});

test("unlisted adapter/surface normalize to vscode", () => {
  const ev = buildMetricsEvent({
    installId: "x",
    vscodeExtensionVersion: "0.0.35",
    adapter: "antigravity",
    surface: "telepathy",
  });
  assert.equal(ev.adapter, "vscode");
  assert.equal(ev.surface, "vscode");
});

test("compression_ratio clamps to [0,1]", () => {
  const ev = buildMetricsEvent({
    installId: "x",
    vscodeExtensionVersion: "0.0.35",
    inputBytes: 10,
    savedBytes: 999,
  });
  assert.ok(ev.compression_ratio <= 1 && ev.compression_ratio >= 0);
});

test("compareVersions handles upgrade detection", () => {
  assert.equal(compareVersions("0.0.34", "0.0.35"), -1);
  assert.equal(compareVersions("0.0.35", "0.0.35"), 0);
  assert.equal(compareVersions("v0.0.36", "0.0.35"), 1);
});

// evaluateVersionGate: the 0.2.51 fix's VS Code-side parity. current==minimum
// must never be "blocked", even when the server declares severity="blocked" —
// the same equality bug class fixed in internal/updatecheck on the CLI. The
// 0.2.51 follow-up also mirrors the CLI's current>=latest rule: nothing to
// upgrade to always evaluates to "info", regardless of server severity.
test("below minimum is blocked", () => {
  const r = evaluateVersionGate("0.0.34", "0.0.35", "0.0.36", "info");
  assert.equal(r.belowMinimum, true);
  assert.equal(r.severity, "blocked");
});

test("0.0.35 < min 0.0.36 -> blocked (spec case)", () => {
  const r = evaluateVersionGate("0.0.35", "0.0.36", "0.0.36", "required");
  assert.equal(r.belowMinimum, true);
  assert.equal(r.severity, "blocked");
});

test("0.0.36 == min == latest -> allowed (spec case)", () => {
  const r = evaluateVersionGate("0.0.36", "0.0.36", "0.0.36", "blocked");
  assert.equal(r.belowMinimum, false);
  assert.equal(r.upgradeNeeded, false);
  assert.equal(r.severity, "info");
});

test("0.0.37 > latest 0.0.36 -> allowed, no update required (spec case)", () => {
  const r = evaluateVersionGate("0.0.37", "0.0.36", "0.0.36", "required");
  assert.equal(r.belowMinimum, false);
  assert.equal(r.upgradeNeeded, false);
  assert.equal(r.severity, "info");
});

test("equal to minimum is allowed, even if server says blocked", () => {
  const r = evaluateVersionGate("0.0.35", "0.0.35", "0.0.36", "blocked");
  assert.equal(r.belowMinimum, false);
  assert.equal(r.severity, "required"); // downgraded from the server's "blocked"
});

test("above minimum is allowed, even if server says blocked", () => {
  const r = evaluateVersionGate("0.0.36", "0.0.35", "0.0.37", "blocked");
  assert.equal(r.belowMinimum, false);
  assert.equal(r.severity, "required");
});

test("server severity below blocked passes through unchanged when not below minimum", () => {
  const r = evaluateVersionGate("0.0.36", "0.0.35", "0.0.37", "recommended");
  assert.equal(r.belowMinimum, false);
  assert.equal(r.severity, "recommended");
});

test("no minimum configured is never below minimum", () => {
  const r = evaluateVersionGate("0.0.30", undefined, "0.0.37", "blocked");
  assert.equal(r.belowMinimum, false);
  assert.equal(r.severity, "required");
});

test("current ahead of latest is always info, even if server says blocked", () => {
  const r = evaluateVersionGate("0.0.38", "0.0.35", "0.0.36", "blocked");
  assert.equal(r.belowMinimum, false);
  assert.equal(r.upgradeNeeded, false);
  assert.equal(r.severity, "info");
});

File: internal/telemetry/telemetry_test.go
package telemetry

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"testing"
	"time"
)

// forbiddenKeys are payload fields that must NEVER appear on the wire.
var forbiddenKeys = []string{
	"raw_output", "raw_log", "prompt", "ai_reply", "command",
	"cwd", "path", "file_name", "repo_name", "username", "email",
	"api_key", "token", "full_session_id", "full_host_instance_hash",
	"full_workspace_hash",
	"channel_id", "run_id", "turn_id", "event_id",
	"full_channel_id", "full_run_id", "full_turn_id",
}

func newTestClient(t *testing.T) *Client {
	t.Helper()
	home := t.TempDir()
	c := NewClient(home, "0.2.49")
	c.now = func() time.Time { return time.Date(2026, 6, 29, 0, 0, 0, 0, time.UTC) }
	return c
}

// TestPayloadHasNoSensitiveFields is the core privacy guarantee: even when a
// caller would love to attach command/cwd/output, the marshalled event cannot
// contain any forbidden key.
func TestPayloadHasNoSensitiveFields(t *testing.T) {
	c := newTestClient(t)
	ev := c.Build(Metrics{
		Event:        "run.finished",
		Adapter:      "claude",
		Surface:      "cli",
		InputBytes:   100000,
		SummaryBytes: 1000,
		SavedBytes:   99000,
		RunCount:     1,
		Status:       "success",
		ErrorKind:    "none",
	})
	data, err := json.Marshal(ev)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	var m map[string]any
	if err := json.Unmarshal(data, &m); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	for _, k := range forbiddenKeys {
		if _, ok := m[k]; ok {
			t.Fatalf("forbidden key %q present in telemetry payload: %s", k, data)
		}
	}
	// Also assert the literal substrings never leak (covers nested/renamed cases).
	lower := strings.ToLower(string(data))
	for _, bad := range []string{"raw_output", "\"prompt\"", "\"command\"", "\"cwd\"", "api_key"} {
		if strings.Contains(lower, bad) {
			t.Fatalf("payload leaked %q: %s", bad, data)
		}
	}
}

func TestAllowedKeysOnly(t *testing.T) {
	c := newTestClient(t)
	ev := c.Build(Metrics{Adapter: "codex", Surface: "hook", InputBytes: 10, SavedBytes: 5})
	data, _ := json.Marshal(ev)
	var m map[string]any
	_ = json.Unmarshal(data, &m)
	allowed := map[string]bool{
		"schema": true, "event": true, "anonymous_install_id": true, "timestamp": true,
		"cli_version": true, "vscode_extension_version": true, "adapter": true, "surface": true,
		"os": true, "arch": true, "input_bytes": true, "summary_bytes": true, "saved_bytes": true,
		"estimated_saved_tokens": true, "compression_ratio": true, "run_count": true,
		"status": true, "error_kind": true,
	}
	for k := range m {
		if !allowed[k] {
			t.Fatalf("unexpected key in payload: %q", k)
		}
	}
}

func TestSavedTokensEstimate(t *testing.T) {
	cases := []struct{ in, want int }{
		{0, 0}, {3, 0}, {4, 1}, {100, 25}, {122222, 30555}, {-10, 0},
	}
	for _, tc := range cases {
		if got := EstimatedSavedTokens(tc.in); got != tc.want {
			t.Fatalf("EstimatedSavedTokens(%d)=%d want %d", tc.in, got, tc.want)
		}
	}
}

func TestCompressionRatio(t *testing.T) {
	if r := CompressionRatio(100000, 92000); r < 0.91 || r > 0.93 {
		t.Fatalf("ratio=%v want ~0.92", r)
	}
	if r := CompressionRatio(0, 10); r != 0 {
		t.Fatalf("ratio with zero input should be 0, got %v", r)
	}
	if r := CompressionRatio(10, 999); r != 1 {
		t.Fatalf("ratio should clamp to 1, got %v", r)
	}
}

func TestAdapterAndSurfaceNormalization(t *testing.T) {
	c := newTestClient(t)
	ev := c.Build(Metrics{Adapter: "antigravity", Surface: "weird"})
	if ev.Adapter != "unknown" {
		t.Fatalf("non-listed adapter should be 'unknown', got %q", ev.Adapter)
	}
	if ev.Surface != "cli" {
		t.Fatalf("non-listed surface should fall back to 'cli', got %q", ev.Surface)
	}
}

// TestCodexSurfaceValuesPassThrough locks in backward-compatible support for
// the finer-grained Codex front-end breakdown (0.2.51): adapter stays
// "codex" for all of them, and each of the new surface values normalizes to
// itself rather than collapsing to the generic "cli" fallback.
func TestCodexSurfaceValuesPassThrough(t *testing.T) {
	c := newTestClient(t)
	for _, surface := range []string{"codex_cli", "codex_ide", "chatgpt_desktop_codex", "codex_shared"} {
		ev := c.Build(Metrics{Adapter: "codex", Surface: surface})
		if ev.Adapter != "codex" {
			t.Fatalf("surface=%s: adapter should stay codex, got %q", surface, ev.Adapter)
		}
		if ev.Surface != surface {
			t.Fatalf("surface=%s: expected pass-through, got %q", surface, ev.Surface)
		}
	}
}

func TestEnabledEnvOverrideHighestPriority(t *testing.T) {
	home := t.TempDir()
	// Persist enabled=true in the file, then force off via env.
	if err := SetEnabled(home, true); err != nil {
		t.Fatal(err)
	}
	t.Setenv("XIT_TELEMETRY", "off")
	if Enabled(home) {
		t.Fatal("XIT_TELEMETRY=off must win over config")
	}
	t.Setenv("XIT_TELEMETRY", "on")
	if !Enabled(home) {
		t.Fatal("XIT_TELEMETRY=on must enable")
	}
}

func TestDefaultEnabled(t *testing.T) {
	home := t.TempDir()
	os.Unsetenv("XIT_TELEMETRY")
	os.Unsetenv("DO_NOT_TRACK")
	if !Enabled(home) {
		t.Fatal("telemetry should be enabled by default (anonymous-by-default)")
	}
}

func TestSetEnabledPersists(t *testing.T) {
	home := t.TempDir()
	os.Unsetenv("XIT_TELEMETRY")
	if err := SetEnabled(home, false); err != nil {
		t.Fatal(err)
	}
	if Enabled(home) {
		t.Fatal("after SetEnabled(false), telemetry should be off")
	}
	en, src := EnabledSource(home)
	if en || !strings.Contains(src, "config") {
		t.Fatalf("expected disabled-from-config, got en=%v src=%q", en, src)
	}
}

func TestEmitDisabledSendsNothing(t *testing.T) {
	home := t.TempDir()
	var hits int
	var mu sync.Mutex
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		mu.Lock()
		hits++
		mu.Unlock()
		w.WriteHeader(200)
	}))
	defer srv.Close()

	t.Setenv("XIT_TELEMETRY", "off")
	t.Setenv("XIT_API_BASE", srv.URL)
	c := NewClient(home, "0.2.49")
	c.Emit(Metrics{Adapter: "cli", Surface: "cli"})
	time.Sleep(100 * time.Millisecond)
	mu.Lock()
	defer mu.Unlock()
	if hits != 0 {
		t.Fatalf("telemetry off must not send, got %d hits", hits)
	}
}

func TestEmitEnabledPostsEvent(t *testing.T) {
	home := t.TempDir()
	got := make(chan []byte, 1)
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		buf := make([]byte, r.ContentLength)
		_, _ = r.Body.Read(buf)
		got <- buf
		w.WriteHeader(202)
	}))
	defer srv.Close()

	t.Setenv("XIT_TELEMETRY", "on")
	t.Setenv("XIT_API_BASE", srv.URL)
	c := NewClient(home, "0.2.49")
	c.Emit(Metrics{Adapter: "codex", Surface: "hook", InputBytes: 100, SavedBytes: 80, Status: "success"})

	select {
	case body := <-got:
		var m map[string]any
		if err := json.Unmarshal(body, &m); err != nil {
			t.Fatalf("server got invalid json: %v (%s)", err, body)
		}
		for _, k := range forbiddenKeys {
			if _, ok := m[k]; ok {
				t.Fatalf("forbidden key %q reached the server", k)
			}
		}
		if m["adapter"] != "codex" {
			t.Fatalf("adapter mismatch: %v", m["adapter"])
		}
	case <-time.After(2 * time.Second):
		t.Fatal("server did not receive telemetry event")
	}
}

func TestEmitFailOpenSpoolsToQueue(t *testing.T) {
	home := t.TempDir()
	t.Setenv("XIT_TELEMETRY", "on")
	t.Setenv("XIT_API_BASE", "http://127.0.0.1:0") // unroutable => delivery fails
	c := NewClient(home, "0.2.49")
	c.deliver(c.Build(Metrics{Adapter: "kimi", Surface: "cli"}))
	q := readQueue(home)
	if len(q) != 1 {
		t.Fatalf("failed delivery should spool 1 event, got %d", len(q))
	}
}

func TestQueueCappedAt100(t *testing.T) {
	home := t.TempDir()
	for i := 0; i < 150; i++ {
		enqueue(home, Event{Schema: SchemaName, Adapter: "cli"})
	}
	if got := len(readQueue(home)); got != maxQueue {
		t.Fatalf("queue should cap at %d, got %d", maxQueue, got)
	}
}

func TestInstallIDStableAndAnonymous(t *testing.T) {
	home := t.TempDir()
	id1 := InstallID(home)
	id2 := InstallID(home)
	if id1 != id2 {
		t.Fatalf("install id should be stable: %q != %q", id1, id2)
	}
	if id1 == "" {
		t.Fatal("install id should not be empty")
	}
	// The id must not embed the username or home path.
	if u := os.Getenv("USER"); u != "" && strings.Contains(id1, u) {
		t.Fatalf("install id leaked username: %q", id1)
	}
	if strings.Contains(id1, filepath.Base(home)) {
		t.Fatalf("install id leaked home path: %q", id1)
	}
}
